kf_lib_data_ingest/validation/data_validator.py

```diff
@@ -0,0 +1,435 @@
+
+                for c, n in nodes:
+                    graph.add_node((c, n))
+
+                # If a node has a direct hierarchical connection within a row, we
+                # can probably assume that indirect connections in the same row are
+                # superfluous
+                #
+                # We should probably also only connect a given start node to its
+                # hierarchically closest destination nodes
+
+                row_edges = list(combinations(nodes, 2))
+                direct = set()
+
+                # Direct edges go in right away and the source nodes get marked
+                for a, b in row_edges:
+                    if b[0] in self.HIERARCHY_DIRECT[a[0]]:
+                        graph.add_edge(a, b)
+                        direct.add(a)
+                    elif a[0] in self.HIERARCHY_DIRECT[b[0]]:
+                        graph.add_edge(b, a)
+                        direct.add(b)
+
+                # Indirect edges get grouped by distance from source node
+                # so we can find the closest destinations for each source
+                row_dict = defaultdict(lambda: defaultdict(deque))
+                for a, b in row_edges:
+                    if (a not in direct) and b[0] in self.HIERARCHY_PATHS[a[0]]:
+                        row_dict[a][self.HIERARCHY_PATHS[a[0]][b[0]]].append(b)
+                    elif (b not in direct) and a[0] in self.HIERARCHY_PATHS[
+                        b[0]
+                    ]:
+                        row_dict[b][self.HIERARCHY_PATHS[b[0]][a[0]]].append(a)
+                    else:
+                        pass
+
+                # Use only the closest indirect destinations for each source
+                for a, cost_bs in row_dict.items():
+                    nearest_bs = sorted(cost_bs.items())[0][1]
+                    for b in nearest_bs:
+                        remaining_edges.add((a, b))
+
+        counts = self._get_node_counts(graph)
+        colwidth = max(map(len, counts.keys()))
+        count_block = [
+            f"\t{k:<{colwidth}} : {v if v > 0 else ''}"
+            for k, v in counts.items()
+        ]
+        count_string = "\n".join(count_block)
+        logger.info(f"{sum(counts.values())} nodes added.\n{count_string}")
+        logger.info(f"{len(graph.edges())} direct edges added.")
+
+        def prune_unneeded(edges):
+            # (A -> B) + (B -> C) doesn't also need (A -> C)
+            if edges:
+                logger.info(
+                    f"{len(edges)} indirect edges left to insert BEFORE pruning."
+                )
+                graph.is_connected.cache_clear()
+                edges = [e for e in edges if not graph.is_connected(e[0], e[1])]
+                logger.info(
+                    f"{len(edges)} indirect edges left to insert AFTER pruning."
+                )
+            return edges or []
+
+        edges = prune_unneeded(remaining_edges)
```

Comment:
Calling this `remaining_edges` (or anywhere else you're saying edge when you really mean pair of nodes in a row) is confusing to me. It seems like it should be called `node_pairs` or something. Or if you want to leave it at `row_edges` then in `prune_unneeded` you should mention that `edges` is a list of node pairs that may or may not be connected by a path in the graph. Its confusing to me to see `graph.is_connected(e[0], e[1])` because I'm thinking huh, why wouldn't the nodes in an edge in a graph be connected 😄

Also, now that I'm thinking about it..could you actually eliminate this method and just add a line above 172 to check if a, b are connected in the graph? If they're not connected, then add them to `remaining_edges`.

Reply:
> Calling this remaining_edges (or anywhere else you're saying edge when you really mean pair of nodes in a row) is confusing to me. It seems like it should be called node_pairs or something. Or if you want to leave it at row_edges then in prune_unneeded you should mention that edges is a list of node pairs that may or may not be connected by a path in the graph. Its confusing to me to see graph.is_connected(e[0], e[1]) because I'm thinking huh, why wouldn't the nodes in an edge in a graph be connected 😄

good point

> Also, now that I'm thinking about it..could you actually eliminate this method and just add a line above 172 to check if a, b are connected in the graph? If they're not connected, then add them to remaining_edges.

Technically I could add lines to this effect earlier, but it would be more complicated than just one line because I'd also want to avoid doing duplicate lookups (which doing it late solves because edges only contains unique pairs) and the same procedure is used elsewhere as well so I'd want this function anyway.